### ISSUE TITLE
makefile fails b/c of excessively long shell statement (incl. explicit use case in wxWidgets)

### ISSUE THREAD

**Issue by jannick0** [vslavik/bakefile#108]
`bakefile` produces makefiles which occasionally generate shell statements exceeding the max shell command length when run. Well noted that when generating makefiles bakefile does not know which shell will be used, hence the max command length is an unknown to `bakefile` at this stage.

Q's:
* Does bakefile have any features to mitigate this issue? 
* For `gcc` linking statement could the list of files to be linked (see culprit example below) be written to a file first and then read by `gcc`?

Here a real life example where `make` tries to feed an excessively long link statement into Windows' `cmd` and fails, since the command is too long and is simply truncated: call [makefile.gcc](https://github.com/wxWidgets/wxWidgets/blob/master/build/msw/makefile.gcc) of [wxWidgets](https://github.com/wxWidgets/wxWidgets) generated by `bakefile` like so:
~~~
mingw32-make ./makefile.gcc MONOLITHIC=1 SHARED=1 BUILD=release
~~~
A *workaround* in this very example is to replace the explicit list of *.o files by a throughly  defined *.o statement in the dll link commands. This approach might not work in a generic case, but here it works by chance. However, it is error prone and needs manual intervention (using a patch) which might fail if `makefile.gcc` is updated by `bakefile`, hence not really efficient.

Many thanks.

**Comment by vadz** [comment]
Unfortunately the development of bakefile 0.x stopped many years ago, so nothing will happen in it (well, unless someone contributes a fix, of course). The only hope of fixing this is to find some workaround in wxWidgets bakefiles but in the (not so?) long term the solution is to stop using bakefile-0.x there entirely...

I'm leaving this open only because I think the problem might actually also exist in bakefile-1.x as well, although I hadn't run into it there yet.

**Comment by jannick0** [comment]
Thanks for the feedback. I spent a bit of time with bakefile v0.2.11 and attached a [patch](https://github.com/vslavik/bakefile/files/2028255/mingw-link-with-in-files.patch.txt) (I am currently unable to push a PR for this) resolving the stated issue. I tested it against the bakefiles of wx'es master and successfully run `makefile.gcc`.

**Idea**: Linking statements can become terribly long because of the large number of object files to be linked into the target (exe, lib, a). This can raise issues as described above, especially if Windows' `cmd` is invoked. Now to shorten the statement `make` writes the object file list to a file which the linker then reads in using the `@file` syntax (cf., e.g., bottom of gcc's [docu page](https://gcc.gnu.org/onlinedocs/gcc/Overall-Options.html)).

The approach can be easily applied in bakefile v1.x. I believe `cmake` does the same.

**Side remark** on the portability of the makefiles generated by bakefile v0.x (I did not look at all at what bakefile v1.x does):
The `if` and `-if` statements to create folders and remove files and folders appear to be the only ones in the makefiles generated by bakefile which make them 'unportable'. A standard trick to overcome that would be to define `RM` and `MD` (for `mkdir`) depending on the used `SHELL`. If `SHELL = cmd`, then you might define
* `RM = >nul del /f/q` (unix-like: `rm -f`)
* `MD = >nul md` (unix-like: `mkdir -p`)
and portably use them like so: `-$(RM)` and `-$(MD)`. 

After manually replacing the`if` statements of wx'es `makefile.gcc` a unix-like `SHELL` could be chosen.

Just in case here the **patch** again:
~~~diff
commit b9945290d4f895de7e04b436bcd469abfe98b169
Date:   Tue May 22 19:47:12 2018 +0200

    makefile_defs_mingw.bkl: escaping make command for xml syntax

diff --git a/rules/makefile_defs_mingw.bkl b/rules/makefile_defs_mingw.bkl
index a3d8484..a0051d0 100644
--- a/rules/makefile_defs_mingw.bkl
+++ b/rules/makefile_defs_mingw.bkl
@@ -66,17 +66,17 @@
     </template>
     <template id="__commands_templ">
         <set var="__LINK_EXE_CMD" eval="0">
-            $(file >$@.in,$(subst \,/,$(__objects_var)))
+            $(DOLLAR)(file &gt;$@.in,$(DOLLAR)(subst &#92;,&#47;,$(__objects_var)))
             $(__linker) -o $@ @$@.in $(__ldflags_all) $(__ldlibs)
         </set>
         <set var="__LINK_LIB_CMD" eval="0">
             if exist $@ del $@
-            $(file >$@.in,$(subst \,/,$(__objects_var)))
+            $(DOLLAR)(file &gt;$@.in,$(DOLLAR)(subst &#92;,&#47;,$(__objects_var)))
             $(AR) $(AROPTIONS) $@ @$@.in
             $(RANLIB) $@
         </set>
         <set var="__LINK_DLL_CMD" eval="0">
-            $(file >$@.in,$(subst \,/,$(__objects_var)))
+            $(DOLLAR)(file &gt;$@.in,$(DOLLAR)(subst &#92;,&#47;,$(__objects_var)))
             $(__dll_linker) $@ @$@.in $(__ldflags_all) $(__ldlibs)
         </set>
         <set var="__LINK_MODULE_CMD" eval="0">

commit 0cfcf548b3fb0cb5a3406609a181512af78e5d7d
Date:   Tue May 22 18:12:55 2018 +0200

    makefile_defs_mingw.bkl: The object file list is written to an .in file
    in linking statements which the linker reads in upon invocation (using
    the '@file' syntax).
    
    Backslashes in the object file paths are converted to forward slashes
    to avoid issues with compilers.
    
    If make is run with SHELL = cmd, the make rule command is written to a
    temporary batch file which is then called to execute the command. However,
    the length of each line of a batch file is subject to the maximum length
    of a cmd command string.
    
    The approach below should work for every shell (incl. unix-like shells).
    
    The .in file is intentionally not removed after command execution.

diff --git a/rules/makefile_defs_mingw.bkl b/rules/makefile_defs_mingw.bkl
index 8ae5f5c..a3d8484 100644
--- a/rules/makefile_defs_mingw.bkl
+++ b/rules/makefile_defs_mingw.bkl
@@ -66,15 +66,18 @@
     </template>
     <template id="__commands_templ">
         <set var="__LINK_EXE_CMD" eval="0">
-            $(__linker) -o $@ $(__objects_var) $(__ldflags_all) $(__ldlibs)
-        </set>        
+            $(file >$@.in,$(subst \,/,$(__objects_var)))
+            $(__linker) -o $@ @$@.in $(__ldflags_all) $(__ldlibs)
+        </set>
         <set var="__LINK_LIB_CMD" eval="0">
             if exist $@ del $@
-            $(AR) $(AROPTIONS) $@ $(__objects_var)
+            $(file >$@.in,$(subst \,/,$(__objects_var)))
+            $(AR) $(AROPTIONS) $@ @$@.in
             $(RANLIB) $@
         </set>
         <set var="__LINK_DLL_CMD" eval="0">
-            $(__dll_linker) $@ $(__objects_var) $(__ldflags_all) $(__ldlibs)
+            $(file >$@.in,$(subst \,/,$(__objects_var)))
+            $(__dll_linker) $@ @$@.in $(__ldflags_all) $(__ldlibs)
         </set>
         <set var="__LINK_MODULE_CMD" eval="0">
             $(__module_linker) $@ $(__objects_var) $(__ldflags_all) $(__ldlibs)
~~~

I hope this helps. Please edit and amend as you may find appropriate to apply the patch (for the bakefile and wx repo).

Thanks.

**Comment by vadz** [comment]
Thanks, I'll have a look at this soon, but I wonder why do you think it's useful to leave `.in` files? I'd definitely remove them, there doesn't seem to be any good reason to clutter the build directory with them...

BTW, I also think that `.in` is not the right extension to use, it's used, by convention, for template files from which other files, without this extension, are generated. Something like `.rsp` ("response") would probably be better. But it's a trivial change, of course.

**Comment by jannick0** [comment]
Thanks. I am not concerned about naming, of course, as long as things are running thru.   However, I am wondering why non-portable syntax is created by `bakefile` (which is one of the reasons as to why I was hesitating to remove the `.in` files). Was there any special reason for that in the days back when the code generator was set up?

wx'es `makfile.gcc`: If the Windows user was not forced to use `cmd` on a machine with a unix-like shell (e.g. b/c cygwin is installed), then the maximum command line length issue would become less of a pain, since for cygwin's `sh` the threshold is far higher than for Windows' `cmd`. Windows `powershell` would be another option, but I did not investigate in this direction.

**Comment by vslavik** [comment]
> However, I am wondering why non-portable syntax is created by

Because you are looking at *GNU toolchain* output. Doesn't exactly make sense to consider "portable" makefiles. Use Autoconf toolchain for maximum portability.

And of course, the `@file` syntax is not portable either.

**Comment by jannick0** [comment]
Given the context of this PR maximum portability was never my point here, but the question as to why the syntax of makefiles generated by `bakefile -f mingw` unneccessarily restricts the SHELL to cmd.  

Anyways, happy to see the raised issue amended - now that a technique to cleanly solve it is offered.

**Comment by vadz** [comment]
Sorry for returning to this so late, but better late than never...

Unfortunately I think we have a problem here: `$(file)` function is in GNU make 4 and later only and I don't think we can drop support for 3.82 or even 3.81 (AFAICS this one was last included in MinGW 4.4) yet. We could, perhaps, use it conditionally, by testing for `MAKE_VERSION` or something in `.FEATURES` specific to 4.0. Or we could do something specifically for the problematic case of the monolithic build and split the object files list in 2 and use `echo` to put it into a response file.

For now I'm leaving your changes and mine in the response-files branch, but I don't think we can really apply them, unfortunately.

**Comment by jannick0** [comment]
I saw you reopened the related wxwidgets.org [ticket](https://trac.wxwidgets.org/ticket/18135#comment:9).

Here an idea to overcome the `$(file)` issue using `$(foreach)` instead which I hope was already around in the ancient 3.x make versions you are referring to above: Use, e.g.,
~~~
  $(foreach f,$(subst \,/,$(__objects_var)),$(shell echo $f >> $@.in.tmp))
  move /y $@.in.tmp $@.in
  $(__linker) -o $@ @$@.in $(__ldflags_all) $(__ldlibs) # unchanged
~~~
instead of
~~~
  $(file >$@.in,$(subst \,/,$(__objects_var)))
  $(__linker) -o $@ @$@.in $(__ldflags_all) $(__ldlibs)
~~~
That would certainly be not hyper fast, but it should work (I did not test). Note that it is important to place the `$(foreach)` statement into the first recipe line to avoid execution precedence issues.

Would that help?

**Comment by vadz** [comment]
This is a nice idea, it's almost certainly going to be quite slow to launch shell so many times, but it should still be much less than the time taken by linking itself, so I think we could live with it. Could you please test it? If it works, we could merge the changes and finally close this ticket.

BTW, why do we need `move` here? Couldn't we just use the temporary file (which should still have `.rsp` rather than `.in` extension IMO, but this is, of course, a detail) directly?

TIA!

**Comment by jannick0** [comment]
> Could you please test it? If it works, we could merge the changes and finally close this ticket.

I'll test and put together a PR, probably on top of your `reponse-files` branch ... let's see.

> BTW, why do we need `move` here? Couldn't we just use the temporary file ...

Safety first - a link statement failure leaves the system with the response file. When running `make` again, the existing response file gets cluttered with duplicates which is unwanted. So using `move` ensures to always start with a fresh response file once the target gets updated.
I was having the idea to remove an existing response file at the beginning of the recipe block, i.e. putting `if exist $@.rsp del $@.rsp` before `$(foreach)`.  But that would not help due to a `$(foreach)` precedence issue for `make`: `make` always executes a `$(foreach)` statement of a recipe statement block **before** the remaining recipe lines regardless its position in the recipe block (see my [make bug report](https://lists.gnu.org/archive/html/bug-make/2019-06/msg00082.html) from today).

BTW: I am not so sure if it is always good to remove the response file as it contains valuable information difficult to retrieve otherwise.  But let's do it for now.

> ... (which should still have `.rsp` rather than `.in` extension IMO, but this is, of course, a detail) directly?

I'll make sure to use `.rsp`.

**Comment by vadz** [comment]
This was fixed by 88f283c2f741bbaf073246b79d91e202505243fc which is part of v0.2.12, thanks again.
